compiler/rustc_hir_analysis/src/check/always_applicable.rs

```diff
@@ -0,0 +1,178 @@
+use rustc_data_structures::fx::FxHashSet;
+use rustc_errors::{struct_span_err, ErrorGuaranteed};
+use rustc_infer::infer::outlives::env::OutlivesEnvironment;
+use rustc_infer::infer::{RegionResolutionError, TyCtxtInferExt};
+use rustc_middle::ty::subst::SubstsRef;
+use rustc_middle::ty::util::CheckRegions;
+use rustc_middle::ty::{self, TyCtxt};
+use rustc_trait_selection::traits::{self, ObligationCtxt};
+
+use crate::hir::def_id::DefId;
+
+/// This function confirms that the trait implementation identified by
+/// `trait_impl_def_id` is not any more specialized than the type it is
+/// attached to (Issue #8142).
+///
+/// This means:
+///
+/// 1. The self type must be nominal (this is already checked during
+///    coherence),
```

Comment:
This is not enforced for negative impls, is it? I don't see any equivalen to `visit_implementation_of_drop`.